codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java

```diff
@@ -265,18 +271,21 @@ public Http2Stream resetSent() {
         }
 
         @Override
-        public final Object setProperty(Object key, Object value) {
-            return data.put(key, value);
+        public final <K, V> V setProperty(PropertyKey<K, V> key, V value) {
+            checkNotNull(key, "key");
+            return properties.add((DefaultPropertyKey<K, V>) key, value);
```

Comment:
I'm wondering if the `PropertyKey` interface should be:

```java
public interface PropertyKey<T> {
  int index();
  T key();
  boolean isValueTypeCompatible(Class<?> valueClass);
}
```

All the algorithms could just use this interface instead of having to worry about a cast.  This may provide a few opportunities:
1. The user may be able to supply "standard" properties that don't have to be registered (return fixed index, etc...). This would help fufil the "static" use case.
2. To avoid overlapping property index we could provide an interface such that the `PropertyKeyRegistry` starts at some number `x` instead of `0`.  This would essentially be "reserving" a chunk of indices to manage for "static" type keys that are the same across all connections.